### ISSUE TITLE
"Could not load file or assembly "Landis.library.UniversalCohorts-v1" when trying to run a test scenario with Biomass Succession or NECN with Docker image of LANDIS-II v8

### ISSUE THREAD

**Issue by Klemet** [CU-ESIIL/ExtremeWildfire#2]
After building a Docker image with the current Dockerfile, and then using the following commands to run a test simulation with Biomass Succession : 

``` shell
git clone https://github.com/LANDIS-II-Foundation/Extension-Biomass-Succession.git
cd /testings/CoreV8.0-BiomassSuccession7.0
console scenario.txt
```

The following error is returned : 

![image001](https://github.com/user-attachments/assets/ed1c23b5-95e2-4c24-a042-6158846a11ac)

The same error comes when trying the same thing with NECN instead of Biomass Succession (e.g. bit using git clone and then launching [this test scenario](https://github.com/LANDIS-II-Foundation/Extension-NECN-Succession/tree/master/testing/Core8-NECN8-Landscape) in their repository).

I get the same issue (although a little bit different) with my own Dockerfile (see [Clean_Docker_LANDIS-II_8_AllExtensions.zip](https://github.com/user-attachments/files/17982328/Clean_Docker_LANDIS-II_8_AllExtensions.zip)).

**Comment by achubaty** [comment]
It's worth noting (per my second question in #1) that the Dockerfile in this repo copies all the .dll files from `Support-Library-Dlls-v8` into the extension directory. I don't see where you do the same in your Dockerfile. Does copying these over to each extension resolve the problem?

**Comment by Klemet** [comment]
Hey @achubaty 😄 !

If you were referring to the Dockerfile I linked, the operation to copy the support libraries is done at lines 62-65 :

```shell
# Downloading support libraries and installing them
RUN cd /bin/LANDIS_Linux && git clone https://github.com/LANDIS-II-Foundation/Support-Library-Dlls-v8.git \
&& mv /bin/LANDIS_Linux/Support-Library-Dlls-v8/* /bin/LANDIS_Linux/build/extensions \
&& rm -r /bin/LANDIS_Linux/Support-Library-Dlls-v8
```

So, not very different (I think) from what is done at [line 52 of the Dockerfile here](https://github.com/CU-ESIIL/ExtremeWildfire/blob/adc3fe800c7033d7a53df8e0abd80ceae7110c30/docker/landis2/Dockerfile#L52C1-L52C146).

**Comment by achubaty** [comment]
Yes, but see e.g., line 60, where the libs are copied to each extension's `src/lib`  directory:

```
cp /opt/Core-Model-v8-LINUX-main/build/extensions/Support-Library-Dlls-v8-main/*.dll /opt/Core-Model-v8-LINUX-main/build/extensions/Extension-Social-Climate-Fire-master/src/lib/
```

**Comment by Klemet** [comment]
Great catch ! You're right; sorry for misunderstanding your point.

It's true that I'm not doing it in this exact way, but I think I'm still using the libraries of the `Support-Library-Dlls-v8` repo in the end during the build :

- I put all the libraries in `/bin/LANDIS_Linux/build/extensions` (see previous comment) at the beginning of the Dockerfile
- Then, for each extension, the script `editing_csproj_LANDIS-II_files.py` changes the hint paths in the `.csproj` file of the extension to point to `/bin/LANDIS_Linux/build/extensions`, where the libraries will be found, by making all hint paths start with `<HintPath>..\\..\\build\\extensions\\`.

So, I don't see where other versions than the libraries from `Support-Library-Dlls-v8` could be used in that process 🤔.

**Comment by Klemet** [comment]
Hey there !

Update : Thomas Estabrook from the University of Michigan seems to have worked out a part of the problem : it seems to be indeed linked to which version of a library is loaded, as several version can exist at once in different places.

We succeeded in writing a Dockerfile that compiles every available LANDIS-II v8 extensions as of today, and they seem to work fine. I'll try to clean it up and test it more once the remaining extensions (especially PnET) become available for v8. I can't say for sure what fixes the problem of the .dll here, but hey, it works 🤷‍♂️ !

[Clean_Docker_LANDIS-II_8_AllExtensions.zip](https://github.com/user-attachments/files/18509933/Clean_Docker_LANDIS-II_8_AllExtensions.zip)

**Comment by wesranch** [comment]
This issue helped a lot! @Klemet I added the DGS succession extension and GIPL and SHAW libraries building off of your Dockerfile. I wanted to ask if you had any trouble starting containers in detached mode? I can only run interactive but I tried to add an entry point script with no luck.

[docker_wr.zip](https://github.com/user-attachments/files/18793348/docker_wr.zip)

**Comment by Klemet** [comment]
Glad it was of help, @wesranch 😁 !

For detached mode : I never used detached mode per se, but I've found that the following command works great to run things in non-interactive mode at the least.

```shell
docker run --mount type=bind,src="<SCENARIO_FOLDER_FULL_PATH_ON_COMPUTER>",dst=/scenarioFolder <LANDIS_DOCKER_IMAGE_NAME> /bin/sh -c "cd /scenarioFolder && dotnet $LANDIS_CONSOLE scenario.txt"
```

The command is explained in the readme of [https://github.com/LANDIS-II-Foundation/Tool-Docker-Apptainer](https://github.com/LANDIS-II-Foundation/Tool-Docker-Apptainer) . Hope it works for you !
